### ISSUE TITLE
Lightmap rendering issue (r69 vs r71) 

### ISSUE THREAD

**Issue by Fabrice3D** [mrdoob/three.js#6573]
As seen on this pict made running 2 exact same script, one running on r69 the other r71. There is an  obvious diff between the two, where r69 does produce the expected looking at the original bitmaps (lightmap and diffuse) in Photoshop.

I'm no shader hero, but replacing this fragment shader line

    //THREE.ShaderChunk[ 'lightmap_fragment'] = "#ifdef USE_LIGHTMAP\n\n	outgoingLight *= diffuseColor.xyz * texture2D( lightMap, vUv2 ).xyz;\n\n#endif";

by this one, fixed it for me.

    THREE.ShaderChunk[ 'lightmap_fragment'] = "#ifdef USE_LIGHTMAP\n\n	outgoingLight =  outgoingLight.xyz * texture2D( lightMap, vUv2 ).xyz;\n\n#endif";

![screen shot 2015-05-18 at 14 32 30](https://cloud.githubusercontent.com/assets/550453/7681445/aa3dc1f2-fd70-11e4-9ebe-4346dbb721c4.jpg)

**Comment by AlexeyGapon** [comment]
Could you share a code example how did you create such nice looking shading with light maps.

**Comment by Fabrice3D** [comment]
For this specific rendering, it's all in the secondary uv set. I have proposed here a fix a few months back for this. The material on itself is a MeshPhongMaterial with a lightMap object. Nothing special.

**Comment by AlexeyGapon** [comment]
I see. I tried to reproduce lightmap example with smooth light shading but with no luck.
Don`t you know some code examples with implementing global illumination shadows for the  webgl. 
I wish Three.js has it.

**Comment by WestLangley** [comment]
@Fabrice3D Lightmaps are now treated as a source of ambient light. The previous implementation was not correct.

**Comment by Fabrice3D** [comment]
@WestLangley  Can you develop a bit? Because testing again with previous shader, I keep getting this result that is way off the expected result.

**Comment by WestLangley** [comment]
Please provide 2 simple live test cases so we can see what you are doing.

See "How to report a bug" in the [guidelines](https://github.com/mrdoob/three.js/blob/master/CONTRIBUTING.md)?

**Comment by Fabrice3D** [comment]
@WestLangley Here you go:
http://closier.nl/webgl/r71/
http://closier.nl/webgl/r69/

**Comment by WestLangley** [comment]
@Fabrice3D Thank you for the simple examples! I wish everyone would do that. : - )

When I said

> Lightmaps are now treated as a source of ambient light.

that comment referred to changes that ocurred in r.72dev. Sorry for the confusion.

It may be that r.71 is not correct either. It is different that r.69, so do not expect the same result.

/ping @bhouston

Also, perhaps unrelated, `webgl_materials_lightmap.html` in the r.72dev branch is completely messed up -- both colors and textures. /ping @mrdoob

**Comment by Fabrice3D** [comment]
I wish everyone would do that
You're welcome :)

I leave you/your team do your color sauce :)  The provided shader fix above fixes my all problems, tested into more complex sceneries where lights and lightmaps are used. Result are fine and as expected.
Using R71 as is, burns everything. Might help someone till a nice r72 sees the light(map)   :)

**Comment by WestLangley** [comment]
@Fabrice3D You are correct. In r.71 the `MeshBasicMaterial` fragment shader has this code:

    outgoingLight = diffuseColor.rgb;

    #ifdef USE_LIGHTMAP

        outgoingLight *= diffuseColor.xyz * texture2D( lightMap, vUv2 ).xyz;

    #endif

which double-counts the diffuse component.

You said you were using `MeshPhongMaterial`, however.

**Comment by WestLangley** [comment]
In r.72dev, `MeshBasicMaterial` no longer supports light maps, presumably because it does not respond to lights. You should be able to use `MeshPhongMaterial` with a light map and no lights, however.

**Comment by Fabrice3D** [comment]
@WestLangley You can see I do use basicMaterial in the examples. And it was obvious to me that some  applied factor was a bit too enthousiastic :)

About r72 changes. I'll see then!

Glad I could contribute and thanks for looking at this closely @WestLangley

**Comment by WestLangley** [comment]
@Fabrice3D It appears to me that your "lightmap" can be equally-considered an "aoMap". In r.72dev,
`Basic` material now supports aoMaps. I have tested your example using `Basic` material with

	aoMap: THREE.ImageUtils.loadTexture( "lightmap.jpg" ),
	aoMapIntensity: 0.25,

and it looks good to me. What do you think?

**Comment by Fabrice3D** [comment]
As long as there is a way to do similar or better. I have no objection on names, really.
I'm more interested to know why you set to 0.25 and not 1 to get the expected result...
Do you have a simple live test? (my turn to ask) :)

**Comment by WestLangley** [comment]
If you don't mind me reusing your model, I'll post something live when I get a chance so you can experiment... 1 was too dark in the dark areas.... Just my opinion.

> I have no objection on names, really.

This is not about names. Lightmaps are different from ambient occlusion maps.

**Comment by WestLangley** [comment]
Well, I've lost it. Sorry. I did something like the following using r.72dev:

	var mat_body = new THREE.MeshBasicMaterial( {

		map: THREE.ImageUtils.loadTexture( "diffuse.jpg" ),

		aoMap: THREE.ImageUtils.loadTexture( "lightmap.jpg" ), // your lightmap used as aoMap
		aoMapIntensity: 0.5,

		side: THREE.DoubleSide,

	});

Remember, aoMaps use the 2nd set of UVs, and your model had them.

**Comment by Fabrice3D** [comment]
@WestLangley thx for the snippet, will try this asap. 
As about using secondary sets, my editor Prefab3D (currently busy on 3.0), exports in binaries as in my posted examples and also json, automatically fixing the prob with shared vertices when using face based geometries constructs and for the json based data (animated or static with precision controls, good for a 50% reductions of filesize). The exports do come with a modified version of the engine for the json, so that secondary uv sets are supported. Without this fix, which I already reported and proposed a fix for, the above code can not be automated as secondary uv sets are ignored by the json parser. Of course json are a zillion time less efficient than binaries, but the other problem I've encountered for exports using binaries is that if you look at the parser of my examples, you'll its using a regular geometry while a buffered one that could be way more efficient, This is because (at least till r71) there is no support for lightmaps or similar for this geometry type. I will find out soon by testing r72, but I hope you guys fix the json issue regarding the secondary sets and allow the buffered geoms to support your above code, which in the end, would allow me to spit out an optional fully automated exports with no code but a load/parse command ideal for static elements and offer a choice for geometry types for atomic exports using binaries.
